content/blog/2020-12-16-compcert/index.md

```diff
@@ -0,0 +1,249 @@
+
+The __sixth pass__ generates a CFG from CminorSel with nodes in [RTL][rtl]; it looks like a [graph of instructions][rtlv]. Included is a size-based heuristic to select which branch of an `if` statement is the fallthrough one. 
+
+Now, we can start optimizing the RTL code. The following passes transform the RTL CFG.
+
+The __seventh pass__ transforms ignored or immediately used function calls into tail calls, which do not need a new stack frame to be allocated. This transformation can remove transitions in the CFG, such as returning a function call result directly instead of a variable assigned to it and used only to return the result. That'll throw a wrench in our proof of correctness, but more on that later. 
+
+The __eighth pass__ performs function inlining based on the presence of the `inline` keyword at the call site. Recursive functions are inlined once, and recursive calls are not inlined (especially since inlining a function that increases the caller's stack block beyond the largest integer causes compilation to fail). Since this pass can eliminate CFG nodes, proving correctness needs to be adjusted.  
+
+The __ninth pass__ renumbers the CFG nodes in post-order with respect to their edges, in order to make the following static analyses more efficient. The algorithm used is Kildall's, outlined in pseudocode [here][kildall]. 
+
+The __tenth pass__ performs constant propagation. This pass may also replace instruction operators in the case that some arguments are statically known. Since the number of transitions may change, correctness needs to take this into account. 
+
+The __eleventh pass__ executes common subexpression elimination, although it's not performed on function calls and it doesn't consider `x + y` and `y + x` equivalent. 
+
+The __twelfth pass__ is dead code elimination. This may make a nonterminating program terminate. For example, suppose a variable is assigned to an integer divided by zero, but that variable is never used. This pass will turn the variable assignment into a `nop` with the justification that the source program was unsound.
+
+The __thirteenth pass__ removes unused global variables. 
+
+The __fourteenth pass__ takes the optimized RTL CFG and transforms it into an [LTL][ltl] CFG, which uses machine registers and stack slots and has basic blocks as nodes instead of individual instructions. Register allocation is done via the [iterated register coalescing algorithm][irca]. The algorithm is implemented in OCaml; if it succeeds, the LTL AST is [verified correct][vregalloc] with respect to the original function.
```

Comment:
Perhaps worth name-checking the term "translation validation" here, which is basically what's going on here.